src/burnchains/affirmation.rs

```diff
@@ -0,0 +1,868 @@
+        self.affirmations.get(reward_cycle as usize).cloned()
+    }
+
+    pub fn push(&mut self, entry: AffirmationMapEntry) {
+        self.affirmations.push(entry)
+    }
+
+    pub fn pop(&mut self) -> Option<AffirmationMapEntry> {
+        self.affirmations.pop()
+    }
+
+    pub fn len(&self) -> usize {
+        self.affirmations.len()
+    }
+
+    pub fn as_slice(&self) -> &[AffirmationMapEntry] {
+        &self.affirmations
+    }
+
+    // used to store to database
+    pub fn encode(&self) -> String {
+        let mut ret = String::with_capacity(self.affirmations.len());
+        write!(&mut ret, "{}", self).expect("BUG: failed to serialize affirmations -- likely OOM");
+        ret
+    }
+
+    // used for database from-row
+    pub fn decode(s: &str) -> Option<AffirmationMap> {
+        if !s.is_ascii() {
+            return None;
+        }
+
+        let mut affirmations = Vec::with_capacity(s.len());
+        for chr in s.chars() {
+            if let Some(next) = AffirmationMapEntry::from_chr(chr) {
+                affirmations.push(next);
+            } else {
+                return None;
+            }
+        }
+        Some(AffirmationMap { affirmations })
+    }
+
+    /// Has `other` diverged from `self`?
+    /// If `other` contains a reward cycle affirmation that is not present in `self`, then yes.
+    /// (Note that this means that if `other` is a prefix of `self`, then no divergence).
+    /// Return the index into `other` where the affirmation differs from `self`.
+    pub fn find_divergence(&self, other: &AffirmationMap) -> Option<u64> {
+        for i in 0..cmp::min(self.len(), other.len()) {
+            if self.affirmations[i] != other.affirmations[i] {
+                return Some(i as u64);
+            }
+        }
+
+        if other.len() > self.len() {
+            return Some(self.len() as u64);
+        }
+
+        None
+    }
+
+    /// What is the PoX ID if this affirmation map?
+    /// This is a surjective mapping: `n` and `p` are 1, and `a` is 0
+    pub fn as_pox_id(&self) -> PoxId {
+        let mut pox_id = PoxId::initial();
+
+        // affirmation maps are statements out prepare phases, not about the reward cycle's anchor
+        // block status.  So, account for the first reward cycle, which has no anchor block.
+        pox_id.extend_with_present_block();
+
+        for affirmation in self.affirmations.iter() {
+            match affirmation {
+                AffirmationMapEntry::PoxAnchorBlockAbsent => {
+                    pox_id.extend_with_not_present_block();
+                }
+                _ => {
+                    pox_id.extend_with_present_block();
+                }
+            }
+        }
+        pox_id
+    }
+
+    /// What is the weight of this affirmation map?
+    /// i.e. how many times did the network either affirm an anchor block, or made no election?
+    pub fn weight(&self) -> u64 {
```

Comment:
What's wrong with the current explanation?

Reply:
Yes, "n" counts as a +1.  This is because a block-commit whose affirmation map is `pn` represents a more-recent measurement of the network's understanding of which anchor blocks are known than a block-commit whose affirmation map is just `p`.  This is because a block-commit whose affirmation map is `pn` must have been processed at a strictly later time, and thus represent a more-recent understanding.  Therefore, its affirmation map is "heavier".